src/hmem.c

```diff
@@ -238,3 +248,60 @@ enum fi_hmem_iface ofi_get_hmem_iface(const void *addr)
 
 	return FI_HMEM_SYSTEM;
 }
+
+int ofi_hmem_host_register(void *ptr, size_t size)
+{
+	enum fi_hmem_iface iface;
+	int ret;
+
+	for (iface = 0; iface < ARRAY_SIZE(hmem_ops); iface++) {
+		if (!hmem_ops[iface].initialized)
+			continue;
+
+		ret = hmem_ops[iface].host_register(ptr, size);
+		if (ret != FI_SUCCESS && ret != -FI_ENOSYS)
+			goto err;
+	}
+
+	return FI_SUCCESS;
+
+err:
+	FI_WARN(&core_prov, FI_LOG_CORE,
+		"Failed to register host memory with hmem iface %s: %s\n",
+		fi_tostr(&iface, FI_TYPE_HMEM_IFACE),
+		fi_strerror(-ret));
+
+	for (iface--; iface >= 0; iface--) {
+		if (!hmem_ops[iface].initialized)
+			continue;
+
+		hmem_ops[iface].host_unregister(ptr);
+	}
+
+	return ret;
+}
+
+int ofi_hmem_host_unregister(void *ptr)
+{
+	enum fi_hmem_iface iface;
+	int ret;
+
+	for (iface = 0; iface < ARRAY_SIZE(hmem_ops); iface++) {
+		if (!hmem_ops[iface].initialized)
+			continue;
+
+		ret = hmem_ops[iface].host_unregister(ptr);
+		if (ret != FI_SUCCESS && ret != -FI_ENOSYS)
+			goto err;
+	}
+
+	return FI_SUCCESS;
+
+err:
+	FI_WARN(&core_prov, FI_LOG_CORE,
+		"Failed to unregister host memory with hmem iface %s: %s\n",
+		fi_tostr(&iface, FI_TYPE_HMEM_IFACE),
+		fi_strerror(-ret));
+
+	return ret;
+}
```

Comment:
Please update the commit message to describe what this is doing.  "Register host memory with device memory" doesn't make sense to me.  What does 'registered' mean in this context?  How do you register memory with memory?

Why should this new API be restricted to host memory?

Reply:
Updated the commit message.